contracts/ExampleToken-v2.cdc

```diff
@@ -0,0 +1,242 @@
+        
+        /// Return information about the vault's type and paths
+        pub fun getTypeInfo(): FungibleToken.VaultInfo {
+            return FungibleToken.VaultInfo(type: self.getType(), VaultStoragePath: self.VaultStoragePath, ReceiverPublicPath: self.ReceiverPublicPath, BalancePublicPath: self.BalancePublicPath)
+        }
+
+        /// Get the balance of the vault
+        pub fun getBalance(): UFix64 {
+            return self.balance
+        }
+
+        /// withdraw
+        ///
+        /// Function that takes an amount as an argument
+        /// and withdraws that amount from the Vault.
+        ///
+        /// It creates a new temporary Vault that is used to hold
+        /// the money that is being transferred. It returns the newly
+        /// created Vault to the context that called so it can be deposited
+        /// elsewhere.
+        ///
+        pub fun withdraw(amount: UFix64): @ExampleToken.Vault{FungibleToken.Vault} {
+            self.balance = self.balance - amount
+            emit TokensWithdrawn(amount: amount, from: self.owner?.address, type: self.getType())
+            return <-create Vault(balance: amount)
+        }
+
+        /// getAcceptedTypes optionally returns a list of vault types that this receiver accepts
+        pub fun getAcceptedTypes(): [Type]? {
+            let typeArray: [Type] = []
+            typeArray.append(Type<@ExampleToken.Vault>())
+            return typeArray
+        }
+
+        /// deposit
+        ///
+        /// Function that takes a Vault object as an argument and adds
+        /// its balance to the balance of the owners Vault.
+        ///
+        /// It is allowed to destroy the sent Vault because the Vault
+        /// was a temporary holder of the tokens. The Vault's balance has
+        /// been consumed and therefore can be destroyed.
+        ///
+        pub fun deposit(from: @AnyResource{FungibleToken.Vault}) {
+            let vault <- from as! @ExampleToken.Vault
+            self.balance = self.balance + vault.balance
+            emit TokensDeposited(amount: vault.balance, to: self.owner?.address, type: self.getType())
+            vault.balance = 0.0
+            destroy vault
+        }
+
+        pub fun transfer(amount: UFix64, recipient: Address) {
+            let transferVault <- self.withdraw(amount: amount)
+
+            // Get the recipient's public account object
+            let recipient = getAccount(recipient)
+
+            // Get a reference to the recipient's Receiver
+            let receiverRef = recipient.getCapability(self.ReceiverPublicPath)
+                .borrow<&{FungibleToken.Receiver}>()
+                ?? panic("Could not borrow receiver reference to the recipient's Vault")
+
+            // Deposit the withdrawn tokens in the recipient's receiver
+            receiverRef.deposit(from: <-transferVault)
+        }
+
+        destroy() {
+            ExampleToken.totalSupply = ExampleToken.totalSupply - self.balance
```

Comment:
Is there a way the FungibleToken interface itself can enforce that this is correct?

Reply:
not that I am aware of. @turbolent do you know if it would be possible to implement post conditions for destructors?

Reply:
I would put there something like:
`if self.balance>0` 
99% of the cases it is 0.

Reply:
What would that accomplish @bluesign ? Maybe I am misunderstanding.

Reply:
flow still thinks registers are updated even update is no-op.

For example, currently FlowToken spams this on every block. ( FlowToken.totalSupply as changed )